engine/src/main/java/org/hibernate/validator/internal/engine/ValidatorImpl.java

```diff
@@ -1461,6 +1460,7 @@ else if ( propertyPathNode.getKey() != null ) {
 
 		metaConstraints.addAll( propertyMetaData.getConstraints() );
 		typeArgumentConstraints.addAll( propertyMetaData.getTypeArgumentsConstraints() );
+		propertyPath.removeLeafNode();
```

Comment:
Previously, the path would already have contained the node for the property of interest. Now we'll add that node when validating the constraint (via that new single `validateMetaConstraint()` method). Hence I'm removing it from the path so we don't end up with having it twice.

One could argue that the new approach is a bit less efficient (as we don't pre-select the relevant constraints), but I preferred that for the sake of less redundant code. We can look into optimizing things later on if needed.

Eventually we also should visit the constraints in order of their location. I.e. first all constraints on Foo#bar, then all on Foo#baz etc. Same for method parameters. This will allow us to do less appendNode()/removeNode() calls, i.e. not do it per constraint but per location. It also should help with that BVAL issue about ordered validation of constraints. I think we are getting to a better base for that.